prov/util/src/util_mem_hooks.c

```diff
@@ -0,0 +1,391 @@
+
+	func_ptr = dlsym(RTLD_DEFAULT, symbol);
+	if (func_ptr == replacement) {
+		(void)dlerror();
+		func_ptr = dlsym(RTLD_NEXT, symbol);
+		if (func_ptr == NULL) {
+			 FI_DBG(&core_prov, FI_LOG_MR,
+				"could not find address of original");
+			return -FI_EOVERFLOW;
+		}
+	}
+
+	return FI_SUCCESS;
+}
+
+int ofi_patcher_patch_symbol(const char *symbol_name, 
+			     uintptr_t replacement, uintptr_t *orig)
+{
+	int ret;
+	struct ofi_patcher_patch* patch = (struct ofi_patcher_patch *)
+					   malloc(sizeof(struct ofi_patcher_patch));
+	if (NULL == patch) {
+		return -FI_EOVERFLOW;
+	}
+	
+	patch->super.patch_symbol = strdup (symbol_name);
+	if (NULL == patch->super.patch_symbol) {
+		free(patch);
+		return -FI_EOVERFLOW;
+	}
+	
+	patch->super.patch_value = replacement;
+	patch->super.patch_restore = (void*) ofi_patcher_remove_patch;
+
+	/* Take lock first to handle a possible race where dlopen() is called
+	 * from another thread and we may end up not patching it.
+	 */
+	fastlock_acquire(&ofi_patcher_module.patch_list_mutex);
+	
+	ret = ofi_patcher_apply_patch(patch);
+	if (FI_SUCCESS != ret) {
+		goto unlock;
+	}
+
+	*orig = ofi_patcher_get_orig(patch->super.patch_symbol ,
+				    (void *) replacement);
+	if (FI_SUCCESS != ret) {
+		goto unlock;
+	}
+	dlist_insert_tail(&ofi_patcher_module.patch_list,
+			  &patch->super.super);
+
+	fastlock_release(&ofi_patcher_module.patch_list_mutex);
+
+unlock:
+	free(patch);
+	fastlock_release(&ofi_patcher_module.patch_list_mutex);
+
+	return ret;
+}
+
+
+/* calling __syscall is preferred on some systems when some
+ * arguments may be 64-bit. it also has the benefit of having
+ * an off_t return type */
+#define memory_patcher_syscall syscall
+
+/*SYS_MMAP*/
+#if defined (SYS_mmap)
+
+static void *(*original_mmap)(void *, size_t, int, int, int, off_t);
+
+static void *_intercept_mmap(void *start, size_t length,
+			     int prot, int flags, int fd, off_t offset)
+{
+	void *result = 0;
+
+	ofi_patcher_handler(start, length);
+
+	if (!original_mmap) {
+		result = (void*)(intptr_t) memory_patcher_syscall(SYS_mmap, start, length,
+								  prot, flags, fd, offset);
+	} else {
+		result = original_mmap (start, length, prot, flags, fd, offset);
+	}
+
+	return result;
+}
+
+static void *intercept_mmap(void *start, size_t
+			    length, int prot, int flags, int fd, off_t offset)
+{
+	void *result = _intercept_mmap (start, length, prot, flags, fd, offset);
+
+	return result;
+}
+
+#endif
+
+/*SYS_MUNMAP*/
+#if defined (SYS_munmap)
+
+static int (*original_munmap) (void *, size_t);
+
+static int _intercept_munmap(void *start, size_t length)
+{
+	int result = 0;
+
+	ofi_patcher_handler(start, length);
+
+	if (!original_munmap) {
+		result = memory_patcher_syscall(SYS_munmap, start, length);
+	} else {
+		result = original_munmap(start, length);
+	}
+
+	return result;
+}
+
+static int intercept_munmap(void *start, size_t length)
+{
+	int result = _intercept_munmap (start, length);
+
+	return result;
+}
```

Comment:
There is a fabtests, fi_unmap_mem, that can be used to verify that the intercepts and mr cache are working.  That program will call mmap/munmap in between transfers in order to change the virtual to physical address translations.